internal/directives/json_updater.go

```diff
@@ -0,0 +1,166 @@
+package directives
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"slices"
+	"strings"
+
+	securejoin "github.com/cyphar/filepath-securejoin"
+	"github.com/tidwall/sjson"
+	"github.com/xeipuuv/gojsonschema"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+)
+
+func init() {
+	builtins.RegisterPromotionStepRunner(newJSONUpdater(), nil)
+}
+
+// jsonUpdater is an implementation of the PromotionStepRunner interface that
+// updates the values of specified keys in a JSON file.
+type jsonUpdater struct {
+	schemaLoader gojsonschema.JSONLoader
+}
+
+// newJSONUpdater returns an implementation of the PromotionStepRunner interface
+// that updates the values of specified keys in a JSON file.
+func newJSONUpdater() PromotionStepRunner {
+	r := &jsonUpdater{}
+	r.schemaLoader = getConfigSchemaLoader(r.Name())
+	return r
+}
+
+// Name implements the PromotionStepRunner interface.
+func (j *jsonUpdater) Name() string {
+	return "json-update"
+}
+
+// RunPromotionStep implements the PromotionStepRunner interface.
+func (j *jsonUpdater) RunPromotionStep(
+	ctx context.Context,
+	stepCtx *PromotionStepContext,
+) (PromotionStepResult, error) {
+	failure := PromotionStepResult{Status: kargoapi.PromotionPhaseErrored}
+
+	if err := j.validate(stepCtx.Config); err != nil {
+		return failure, err
+	}
+
+	// Convert the configuration into a typed struct
+	cfg, err := ConfigToStruct[JSONUpdateConfig](stepCtx.Config)
+	if err != nil {
+		return failure, fmt.Errorf("could not convert config into %s config: %w", j.Name(), err)
+	}
+
+	return j.runPromotionStep(ctx, stepCtx, cfg)
+}
+
+// validate validates jsonUpdater configuration against a JSON schema.
+func (j *jsonUpdater) validate(cfg Config) error {
+	return validate(j.schemaLoader, gojsonschema.NewGoLoader(cfg), j.Name())
+}
+
+func (j *jsonUpdater) runPromotionStep(
+	_ context.Context,
+	stepCtx *PromotionStepContext,
+	cfg JSONUpdateConfig,
+) (PromotionStepResult, error) {
+	updates := make(map[string]any, len(cfg.Updates))
+	for _, update := range cfg.Updates {
+		var value any
+		if update.Value != nil {
+			switch {
+			case update.Value.String != "":
+				value = update.Value.String
+			case len(update.Value.AnythingArray) > 0:
+				value = update.Value.AnythingArray
+			case len(update.Value.AnythingMap) > 0:
+				value = update.Value.AnythingMap
```

Comment:
I think I would consider initially supporting only scalar values -- strings, numbers, booleans, and nulls. I think that with arrays and objects/maps in the picture, "there be dragons here." Among other potential problems is the difficulty posed in distilling such changes into a commit message -- which I see you're currently not doing.

Another problem with supporting non-scalar values is that although we can preserve formatting choices in the existing file, we won't be able to make the modifications _match_ the formatting choices of the existing file.

i.e. If you start with:

```
{
  "name": "Bart",
  "address": {
    "street": "42 Evergreen Terrace",
    "city": "Springfield"
  }
}
```

And you update address, you could end up with something like this:

```
{
  "name": "Bart",
  "address": { "street": "123 Fake Street", "city": "Springfield" }
}
```

It's not invalid, but I can imagine users being annoyed with it.

Altogether, I think it's better to start off with what we know we can do well and look to improve it over time if there's a demonstrated need for it.

I think just supporting scalars, would address the vast majority user cases.